File: java/src/main/java/br/com/emendes/problems/BinaryTreeInorderTraversal.java
package br.com.emendes.problems;

import br.com.emendes.problems.util.TreeNode;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * Dado o root de uma árvore binária, retornar a passagem em ordem dos valores de seus nós.<br>
 * <br>
 * Restrições:<br>
 * O número de node da árvore está no range de [0, 100]<br>
 * -100 <= Node.val <= 100<br>
 * <br>
 * Follow up: A solução recursiva é trivial, você poderia fazer isso iterativamente?
 */
public class BinaryTreeInorderTraversal {

  public List<Integer> inorderTraversal(TreeNode root) {
    Deque<TreeNode> stack = new LinkedList<>();
    List<Integer> answer = new ArrayList<>();
    TreeNode node = root;

    while (node != null || !stack.isEmpty()) {
      if (node == null) {
        node = stack.pop();
        answer.add(node.val);
        node = node.right;
      } else {
        stack.push(node);
        node = node.left;
      }
    }

    return answer;
  }

  public List<Integer> firstSolution(TreeNode root) {
    ArrayList<Integer> answer = new ArrayList<>();
    inorderTraversal(root, answer);

    return answer;
  }

  private void inorderTraversal(TreeNode node, List<Integer> answer) {
    if (node != null) {
      inorderTraversal(node.left, answer);
      answer.add(node.val);
      inorderTraversal(node.right, answer);
    }
  }

}

File: javascript/problems/BinaryTreeInorderTraversal.js
/**
 * Dado o root de uma binary tree, retorne a inorder traversal dos valores do nodes.
 *
 * Restrições:
 *
 * - O número de nodes na árvore está no intervalo [0, 100]
 *
 * - -100 <= Node.val <= 100
 *
 * Follow-up: A solução recursiva é trivial. Você poderia fazê-la iterativamente?
 */
const inorderTraversal = function (root) {
  const output = [];
  traverse(root, output);

  return output;
};

const traverse = (node, output) => {
  if (node === null) return;
  traverse(node.left, output);
  output.push(node.val);
  traverse(node.right, output);
};

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}
